docs/src/proposals/partitioned-inflationary-rewards-distribution.md

```diff
@@ -0,0 +1,136 @@
+---
+title: Partitioned Inflationary Rewards Distribution
+---
+
+## Problem
+
+With the increase of number of stake accounts, computing and redeeming the stake
+rewards at the start block of the epoch boundary becomes very expensive.
+Currently, with 550K stake accounts, the stake reward time has already taken
+more than 10 seconds. This prolonged computation slows down the network, and can
+cause large number of forks at the epoch boundary, which makes the matter even
+worse.
+
+## Proposed Solutions
+
+Instead of computing and reward stake accounts at epoch boundary, we will
+decouple reward computation and reward credit into two phases.
+
+A separate service, "epoch calc service" will be created. The service will
+listen to a channel for any incoming rewards calculation requests, and perform
+the calculation for the rewards. For each block that cross the epoch boundary,
+the bank will send a request to the `epoch calc service`. This marks the start
+of the reward computation phase.
+
+N-1 -- N -- N+1
+     \
+      \
+        N+2
+
+In the above example, N is the start of the new epoch. Two rewards calculation
+requests will be sent out at slot N and slot N+2. To avoid repeated computation
+with the same input, the signature of the computation requests,
+hash(epoch_number, hash(stake_accounts_data)), are calculated. Duplicated
+computation requests will be discard. For the above example, if there are no
+stake account changes between slot N-1 and slot N, the 2nd computation request
+will be discarded.
+
+When reaching block height `N` after the start of the `reward computation
+phase`, the bank starts the second phase - reward credit, in which, the bank
+first query the `epoch calc service` with the request signature to get the
+rewards result, then credit the rewards to the stake accounts for the next `M`
+blocks. If the rewards result is not available, the bank will wait until the
+results are available.
+
+We call them:
+(a) `calculating interval` `[epoch_start, epoch_start+N]`
+(b) `credit interval` `[epoch_start+N+1, epoch_start+N+M]`, respectively.
+And the combined interval `[epoch_start, epoch_start+N+M]` is called
+`rewarding interval`.
+
+For `calculating interval`, `N` is chosen to be sufficiently large so that the
+background computation should have completed and the result of the reward
+computation is available at the end of `calculating interval`. `N` can be fixed
+such as 100 (roughly equivalent to 50 seconds), or chosen as a function of the
+number of stake accounts, `f(num_stake_accounts)`.
+
+In `credit interval`,  the bank will fetch the reward computation results from
+the background thread and start credit the rewards during the next `M` blocks.
```

Comment:
I'm a bit confused about the data flow. Will the background thread start writing to accountsdb directly when it finishes computing the result? or it will queue up the writes into memory and then make them to accountsdb when the required bank is created?

Reply:
the easiest implementation is to queue up the bank.store_accounts() calls until the bank is ready. There are other ways I've thought about speeding this up later. But, if we break the stores into enough slots, then we're ok with the default code path.

Reply:
No, the background thread will just compute the rewards for all the accounts and store rewards result into a map, account_pubkey->rewards. And the following M banks each writes 1/M stake accounts.